jzaia_files/main.cpp

```diff
@@ -0,0 +1,291 @@
+    std::iota(controls.begin(), controls.end(), 0);
+    std::vector<bool> control_vals = ORACLE2_EXPECTED;
+
+    // Apply the X gate to the ancilla, controlled on the chosen bitstring
+    GateImplementationsLM::applyNCPauliX(sv.getData(),
+                                         sv.getNumQubits(),
+                                         controls,
+                                         control_vals,
+                                         {ORACLE2_QUBITS-1},
+                                         false);
+}
+
+/**
+ * @brief The third testing oracle for Grover's
+ *
+ * A 17-qubit test oracle for Grover's algorithm. Applies a pauli-X to the
+ * rightmost qubit if the leftmost 16 qubits are in the state |0011001100110011>
+ *
+ * @param sv The statevector to apply the oracle to. Must be 17 qubits
+ */
+void oracle3(StateVectorLQubitManaged<double> &sv) {
+    // Sanity check statevector
+    assert(sv.getNumQubits() == ORACLE3_QUBITS);
+
+    // Define controls to be used for applying the X gate
+    std::vector<size_t> controls(ORACLE3_QUBITS-1);
+    std::iota(controls.begin(), controls.end(), 0);
+    std::vector<bool> control_vals = ORACLE3_EXPECTED;
+
+    // Apply the X gate to the ancilla, controlled on the chosen bitstring
+    GateImplementationsLM::applyNCPauliX(sv.getData(),
+                                         sv.getNumQubits(),
+                                         controls,
+                                         control_vals,
+                                         {ORACLE3_QUBITS-1},
+                                         false);
+}
+
+
+/**
+ * @brief Overall function for running Grover's algorithm on a chosen oracle
+ *
+ * Run Grover's algorithm from start to finish. Prepares a statevector, and
+ * repeats state selection and amplitude-amplification for sqrt(N) iterations
+ * (where N = 2^(# of non-ancilla qubits)). This implementation assumes that
+ * the oracle always picks precisly 1 state (rather than an arbitrary number).
+ *
+ * @param oracle A black-box function that acts on a created statevector
+ * @param num_qubits The number of qubits in the circuit the oracle acts
+ *        on (includes the ancilla)
+ */
+void run_experiment(void (*oracle) (StateVectorLQubitManaged<double> &),
+                    const size_t num_qubits) {
+    // Create the statevector for this circuit
+    // NOTE: qubits are numbered left-to-right order
+    StateVectorLQubitManaged<double> sv(num_qubits);
+
+    // Set up the circuit by placing into uniform superposition
+    groversSetup(sv);
+
+    // Apply (approx) sqrt(N) repetitions of state selection and amp-amp
+    const size_t nreps = static_cast<size_t>(sqrt(1llu << (num_qubits-1)));
+    for (size_t reps = nreps; reps > 0; --reps) {
+        // Apply the oracle to apply a phase of -1 to desired state
+        oracle(sv);
+        // Perform amp-amp by reflecting over |+++...+>
+        groversMirror(sv);
+    }
```

Comment:
Is it possible to rewrite this with STL function(s)? What are the implications?

Reply:
To be honest, I'm not as familiar with C++ as I am with plain C, so I had to look into some STL options. It seems like `std::for_each` or `std::for_each_n` might be good to use here, but there isn't a natural collection to loop over. There might be a readability benefit to something like this, or an immutability guarantee for something with more functional style (although in this case, we are using these functions for their side effects). But honestly, I would be very surprised if there is anything that gets a speed advantage over a plain for loop for something like this, compiler optimization should be very good for this.